Global styles: close stylebook if the editor canvas container slot is not filled

ramonjd opened the pull request:
!! t.


## What?
An enhancement to to https://github.com/WordPress/gutenberg/pull/50081 which reinstates the original method of checking for a slot fill to see if we need to render a custom header title for the editor canvas slot.

It also checks the canvasMode so we can reset the stylebook slot.

## Why and how?

Previously we could check that the stylebook was open by using `useSlotFills` to see if the stylebook slot was filled.

Since changing over to a private slot fill in https://github.com/WordPress/gutenberg/pull/49973, we now have to get the key using `privateKey`.

```js
const { createPrivateSlotFill } = unlock( componentsPrivateApis );
const SLOT_FILL_NAME = 'EditSiteEditorCanvasContainerSlot';
const {
	privateKey,
	Slot: EditorCanvasContainerSlot,
	Fill: EditorCanvasContainerFill,
} = createPrivateSlotFill( SLOT_FILL_NAME );

...

function useHasEditorCanvasContainer() {
	const fills = useSlotFills( privateKey );
	return !! fills?.length;
}
```

## Testing Instructions
1. Open the site editor
2. Open the global styles sidebar. 
3. Check that you can open and use the stylebook and that the editor header title says "Style book"
4. Verify that you can close it via the eye icon. The header should now show the title of the current template being edited.
5. Now reopen the stylebook and close the global styles sidebar
6. The header should now show the title of the current template being edited.


## Screenshots or screencast <!-- if applicable -->

![2023-04-27 12 22 22](https://user-images.githubusercontent.com/6458278/234743595-e00062cf-f97d-4087-9077-cd1f689570b9.gif)

ramonjd committed:
Alternative to https://github.com/WordPress/gutenberg/pull/50081 which uses the original method of checking for a slot fill to see if we need to render a custom header title for the editor canvas slot.

ramonjd committed:
Revert changes in https://github.com/WordPress/gutenberg/pull/50081

ramonjd commented on packages/edit-site/src/components/sidebar-edit-mode/global-styles-sidebar.js:
```diff
@@ -18,35 +17,25 @@ import { GlobalStylesMenuSlot } from '../global-styles/ui';
-		[]
-	);
-
+	const { editorMode, editorCanvasView } = useSelect( ( select ) => {
```
This change reinstates the file before https://github.com/WordPress/gutenberg/pull/50081 was merged given that this PR uses @noisysocks's original method of checking for the existence of the slot fill to determine whether to display the header.

andrewserong reviewed (commented):
Thanks for following up @ramonjd!

In testing, I noticed one slight difference between this PR and the fix in https://github.com/WordPress/gutenberg/pull/50081 to do with whether or not the style book remains closed after closing and re-opening the global styles sidebar. On `trunk`, if we use either of the close buttons for the global styles sidebar, then the state for the style book being closed is preserved. However with this PR, it now remembers whether the style book was opened when re-opening the global styles sidebar:

| Trunk | This PR |
| --- | --- |
| ![2023-04-27 10 09 31](https://user-images.githubusercontent.com/14988353/234728763-0dad35a1-bed9-4296-971e-040bc4afad55.gif) | ![2023-04-27 10 11 02](https://user-images.githubusercontent.com/14988353/234728839-80a3bc73-14e5-4bd4-8e32-31aaac27059a.gif) |

In this case, as a user, it could be expected that closing the global styles sidebar would reset its state, so that reopening it again would get you back to the normal view. What do you think?

noisysocks reviewed (approved):
Works when I test locally 👍 and obviously I like the approach since it's close to what I wrote.

ramonjd commented:
Thanks for testing folks! 

> In this case, as a user, it could be expected that closing the global styles sidebar would reset its state, so that reopening it again would get you back to the normal view. What do you think?

That's a good point 🤔 I just double checked on an older branch and what you've described was the state of the style book before I went ahead and started making changes 😄 so this PR reinstates that behaviour. 

But now you've mentioned it, I personally would kind of expect the style book to go ahead and close itself if the global styles bar was closed. What do you folks think?

Similar to the way that, when I drill down into **_Global styles > Colors > Palettes_** for example, and then toggle the sidebar I'm brought back to the main screen of global styles.

For the style book and the upcoming [revisions UI](https://github.com/WordPress/gutenberg/pull/50089), it does make sense to reset both when the global styles panel is closed, so what's on trunk will work for a while until a new fill is created that doesn't care. I'm wondering if there's a better way to go about it 🤔

noisysocks commented:
I think it makes sense to reset, yeah.

andrewserong commented:
> But now you've mentioned it, I personally would kind of expect the style book to go ahead and close itself if the global styles bar was closed. What do you folks think?

Same, very interesting that it worked the other way around before your fix yesterday, though! It's funny how these things often only become clearer once we're testing them in detail.

I quite like the reset state behaviour, but don't feel too strongly either way.

> For the style book and the upcoming https://github.com/WordPress/gutenberg/pull/50089, it does make sense to reset both when the global styles panel is closed, so what's on trunk will work for a while until a new fill is created that doesn't care. I'm wondering if there's a better way to go about it 🤔

Perhaps it'll become clearer once we have a case where we don't want the state to be cleared? 🤔

ramonjd commented:
> Perhaps it'll become clearer once we have a case where we don't want the state to be cleared?

Very good point, which is a case for leaving it for now. I'll see if I can abstract it a bit in this PR.

Which also leads me to the observation that I think we should also reset when `'edit' !== getCanvasMode()`

ramonjd committed:
Reinstates trunk logic to check for global styles sidebar state AND throws in a check for canvas mode === `edit`, both of which will affect whether we reset the editor canvas container.
If the global sidebar is closed or the canvas is no longer in edit mode, turn off the fills.

ramonjd commented:
@noisysocks and @andrewserong thank you for spending time on this one.

I've gone with a balance: 

1. Reinstated the fill check for the header. Seems like a good utility to hang onto
2. Added an extra check for `'edit' === canvasMode` in order to close the stylebook when NOT in edit mode

I'm overthinking this of course 😄 But it works.

andrewserong reviewed (approved):
Thanks for following up, this looks like a good balance to me!

> Added an extra check for 'edit' === canvasMode in order to close the stylebook when NOT in edit mode

I like that inclusion, too, as it means clicking the W icon escapes the user out of the style book, too. Nice that we're prevent folks from getting stuck!

LGTM ✨
